src/dialog/dlgreplacecuecolor.cpp

```diff
@@ -0,0 +1,382 @@
+    VERIFY_OR_DEBUG_ASSERT(newColor) {
+        m_bDatabaseChangeInProgress = false;
+        slotUpdateApplyButton();
+        return;
+    }
+
+    int hotcueIndex = spinBoxHotcueIndex->value() - 1;
+
+    QSqlDatabase database = mixxx::DbConnectionPooled(m_pDbConnectionPool);
+
+    // Open the database connection in this thread.
+    VERIFY_OR_DEBUG_ASSERT(database.isOpen()) {
+        qWarning() << "Failed to open database for cue color replace dialog."
+                   << database.lastError();
+        m_bDatabaseChangeInProgress = false;
+        slotUpdateApplyButton();
+        return;
+    }
+
+    // Build SELECT query string
+    QMap<QString, QVariant> queryValues;
+    QStringList queryStringConditions;
+    if (conditions.testFlag(ConditionFlag::CurrentColorCheck)) {
+        queryStringConditions << QString(
+                QStringLiteral("color") +
+                (conditions.testFlag(ConditionFlag::CurrentColorNotEqual) ? QStringLiteral("!=") : QStringLiteral("=")) +
+                QStringLiteral(":current_color"));
+        queryValues.insert(QStringLiteral(":current_color"), mixxx::RgbColor::toQVariant(currentColor));
+    }
+    if (conditions.testFlag(ConditionFlag::HotcueIndexCheck)) {
+        queryStringConditions << QString(
+                QStringLiteral("hotcue") +
+                (conditions.testFlag(ConditionFlag::HotcueIndexNotEqual) ? QStringLiteral("!=") : QStringLiteral("=")) +
+                QStringLiteral(":hotcue"));
+        queryValues.insert(QStringLiteral(":hotcue"), QVariant(hotcueIndex));
+    }
+
+    QString queryString = QStringLiteral("SELECT id, track_id, color FROM " CUE_TABLE);
+    if (!queryStringConditions.isEmpty()) {
+        queryString += QStringLiteral(" WHERE ") + queryStringConditions.join(QStringLiteral(" AND "));
+    }
+
+    // Select affected queues
+    QSqlQuery selectQuery(database);
+    selectQuery.prepare(queryString);
+    for (auto i = queryValues.constBegin(); i != queryValues.constEnd(); i++) {
+        selectQuery.bindValue(i.key(), i.value());
+    }
+
+    // Flush cached tracks to database
+    QSet<TrackId> cachedTrackIds = GlobalTrackCacheLocker().getCachedTrackIds();
+    for (const TrackId& trackId : cachedTrackIds) {
+        TrackPointer pTrack = GlobalTrackCacheLocker().lookupTrackById(trackId);
+        if (pTrack) {
+            m_pTrackCollectionManager->saveTrack(pTrack);
+        }
+    }
+
+    if (!selectQuery.exec()) {
+        LOG_FAILED_QUERY(selectQuery);
+        m_bDatabaseChangeInProgress = false;
+        slotUpdateApplyButton();
+        return;
+    }
+    int idColumn = selectQuery.record().indexOf("id");
+    int trackIdColumn = selectQuery.record().indexOf("track_id");
+    int colorColumn = selectQuery.record().indexOf("color");
+
+    QList<CueDatabaseRow> rows;
+    QSet<TrackId> trackIds;
+    while (selectQuery.next()) {
+        QCoreApplication::processEvents();
```

Comment:
This is a bad idea, because every other GUI slot can be executed in between. Also other DB actions or even another color change attempt. The same below.

Reply:
This was suggested by @uklotzde. Also, the dialog is modal so the impact is probably neglible. As a user, if I mess around with colors while the color update is happening I'd actually not be surprised if my color gets overwritten.

Reply:
We really must not do this. This is one of the Qts anti patterns. We had some of these "solutions" in Mixxx and have removed them all.  What is the issue you are trying to solve? There must be a better solution.

Reply:
My initial attempt was doing the DB operations in another thread, but this can cause inconsistencies. Now I moved everything into the GUI thread. Maybe @uklotzde can elaborate why we need this, but I think the GUI would freeze and the user has no possibility to hit the cancel button without this.

Reply:
We have other places in Mixxx where the GUI Freezes more badly because of all kind of Ctrl+A actions. This is nothing we can solve here. Safety over performance. Please just remove this call if we don't have a looking strategy. 
I think we can effort a gui freeze here, because it is only a one time action.

Reply:
I disagree with @daschuer, see my comments (review comments appear at different places in GitHub)

@Holzhaus Please test what happens if you remove the processEvents() call. If the progress dialog becomes unresponsive then removing it is not an option imho.

Reply:
My test database it not big enough to test this.

Reply:
A modal QProgressDialog does not require to call processEvents(). Please remove it, problem solved.

Reply:
The call is redundant but still happens behind the scenes:

> Warning: If the progress dialog is modal (see QProgressDialog::QProgressDialog()), setValue() calls QCoreApplication::processEvents(), so take care that this does not cause undesirable re-entrancy in your code. For example, don't use a QProgressDialog inside a paintEvent()!

Reply:
But as already discussed this is true for this simple dialog.